zilliqa/src/bin/validator_oracle.rs

```diff
@@ -0,0 +1,238 @@
+        let mut chain_clients = vec![Self::create_zq2_chain_client(config.clone(), &wallet).await?];
+        for chain_config in config.chain_configs {
+            chain_clients.push(
+                ChainClient::new(
+                    &chain_config,
+                    H160(**config.zq2.validator_manager_address),
+                    wallet.clone(),
+                )
+                .await?,
+            );
+        }
+
+        Ok(Self {
+            wallet,
+            zq2_config: config.zq2,
+            validators: vec![],
+            chain_clients,
+        })
+    }
+
+    pub async fn start(&mut self) -> Result<()> {
+        info!(
+            "Starting validator oracle with wallet address {}",
+            self.wallet.address()
+        );
+
+        self.validators = self.get_stakers().await?;
+        info!("Validators: {}\n", serde_json::to_string(&self.validators)?);
+
+        let (sender, receiver) = watch::channel(self.validators.clone());
+
+        let mut handles = vec![];
+        for chain_client in &self.chain_clients {
+            let mut receiver = receiver.clone();
+            let chain_client = chain_client.clone();
+            let handle = tokio::spawn(async move {
+                loop {
+                    let validators = receiver.borrow_and_update().clone();
+                    if Self::update_validator_manager(&chain_client, &validators)
+                        .await
+                        .is_err()
+                    {
+                        error!(
+                            "Failed updating the validator manager on {}",
+                            chain_client.rpc_url
+                        );
+                    }
+
+                    if receiver.changed().await.is_err() {
+                        break;
+                    }
+                }
+            });
+            handles.push(handle);
+        }
+
+        let result = self.listen_to_staker_updates(sender).await;
+
+        for handle in handles {
+            handle.await?;
+        }
+
+        result
+    }
+
+    async fn listen_to_staker_updates(
+        &mut self,
+        sender: watch::Sender<Vec<NodePublicKey>>,
+    ) -> Result<()> {
+        let filter = Filter::new()
+            .address(contract_addr::DEPOSIT)
+            // Must be the same signature as the event in deposit.sol
+            .event("StakerAdded(bytes)")
+            .from_block(BlockNumberOrTag::Finalized);
+
+        let ws = WsConnect::new(&self.zq2_config.rpc_url);
+        let provider = ProviderBuilder::new().on_ws(ws).await?;
+
+        let subscription = provider.subscribe_logs(&filter).await?;
+        let mut stream = subscription.into_stream();
+        while let Some(log) = stream.next().await {
+            // TODO: infer if staker added or removed
+            info!("Received validator update: {log:?}");
+
+            self.validators = self.get_stakers().await?;
```

Comment:
Why do we need to call `get_stakers` if we're listening to added and removed logs?

Reply:
Initially the idea was to maintain the list internally (see the previous comment you had and the explanation) and figure out from the event logs which was added and which removed. Ended up being easier just retrieving them all. Possibly later we can main the list internally instead of just calling this here.